csharp/src/Apache.Arrow.Adbc/Core/AdbcStatement.cs

```diff
@@ -0,0 +1,276 @@
+            set { throw new NotImplementedException(); }
+        }
+
+        public virtual void Bind()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Executes the statement and returns a tuple containing the number of records and the <see cref="IArrowArrayStream"/>..
+        /// </summary>
+        /// <returns>A <see cref="ValueTuple"/> where the first item is the number of records and the second is the <see cref="IArrowArrayStream"/>.</returns>
+        public abstract QueryResult ExecuteQuery();
+
+        /// <summary>
+        /// Executes the statement and returns a tuple containing the number of records and the <see cref="IArrowArrayStream"/>..
+        /// </summary>
+        /// <returns>A <see cref="ValueTuple"/> where the first item is the number of records and the second is the <see cref="IArrowArrayStream"/>.</returns>
+        public virtual async ValueTask<QueryResult> ExecuteQueryAsync()
+        {
+            return await Task.Run(() => ExecuteQuery());
+        }
+
+        /// <summary>
+        /// Executes an update command and returns the number of records effected.
+        /// </summary>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public abstract UpdateResult ExecuteUpdate();
+
+        // <summary>
+        /// Executes an update command and returns the number of records effected.
+        /// </summary>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public virtual async Task<UpdateResult> ExecuteUpdateAsync()
+        {
+            return await Task.Run(() => ExecuteUpdate());
+        }
+
+        /// <summary>
+        /// Timeout (in seconds) for statement execution.
+        /// </summary>
+        /// <remarks>The default is 30 seconds.</remarks>
+        public virtual int Timeout { get; set; }
+
+        /// <summary>
+        /// Execute a result set-generating query and get a list of partitions of the result set.
+        /// </summary>
+        /// <returns><see cref="PartitionedResult"/></returns>
+        public virtual PartitionedResult ExecutePartitioned()
+        {
+            throw AdbcException.NotImplemented("Statement does not support executePartitioned");
+        }
+
+        public virtual Schema GetParameterSchema()
+        {
+            throw AdbcException.NotImplemented("Statement does not support GetParameterSchema");
+        }
+
+        public virtual void Prepare()
+        {
+            throw AdbcException.NotImplemented("Statement does not support Prepare");
+        }
+
+        public virtual void Dispose()
+        {
+        }
+
+        /// <summary>
+        /// Gets the .NET type based on the Arrow field metadata
+        /// </summary>
+        /// <param name="f"></param>
+        /// <returns></returns>
+        public virtual Type ConvertArrowType(Field f)

```

Comment:
It seems the below helpers should go somewhere else? Possibly even upstream into the Arrow library itself?

Reply:
I put it there so that each driver can provide its own implementation. For example, in Snowflake, if a Field.DataType.TypeId is ArrowTypeId.Int8, that seems like it would be an sbyte in C#. However, for Snowflake in particular, the Field metadata needs to be examined to determine if the logical type is a _FIXED_. If so, the _scale_ metadata is also examined. If the scale is zero, then the correct type is _long_. However, if scale != 0, then a _decimal_ type is returned.

Reply:
If the raw data is of the wrong type, it should be the driver's responsibility to convert it into the right Arrow type, right? The Go Snowflake driver, for instance, casts the Snowflake result data to the right type after inspecting the type metadata, without pushing this onto the user or into the interface.

Reply:
right -- similar to https://github.com/apache/arrow-adbc/blob/main/go/adbc/driver/snowflake/statement.go